src/metrics/function_error.py

```diff
@@ -0,0 +1,358 @@
+        self.fn_cls = function_class
+        self.num_batches=num_batches
+        self.save_path=save_path
+
+    def _metric(self, ground_truth: Tensor, predictions: Tensor) -> Tensor:
+        """Compute a metric between a prediction and a "ground truth" """
+        raise NotImplementedError("Abstract class FunctionClassError does not implement a metric!")
+
+    def PostProcessingStats(self, errs, model_names, prefix="", B=1000, confidence_level =[0.01,0.05]):
+        stats={} 
+        if (len(prefix)>0):
+            prefix += "_"
+
+        num_models, samples, sequence_length=errs.size()
+        
+        #Bootstrapping 
+        sample_indices = torch.randint(0, samples, (B, samples)) 
+        bootstrap_samples = errs[:,sample_indices,:]
+        means = bootstrap_samples.mean(dim=2)
+        variance_estimate = means.var(dim=1)
+
+
+        std=torch.std(errs, dim=1)
+        mean =torch.mean(errs, dim=1)
+        quantiles=torch.quantile(errs, torch.Tensor([0, 0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99, 1]), dim=1)
+        for i, name in enumerate(model_names):
+            stats[prefix+"accuracy_"+name]=mean[i]
+            stats[prefix+"std_"+name]= std[i]
+            stats[prefix+"std_mean_"+name]= std[i]/np.sqrt(samples)
+            stats[prefix+"max_"+name]=quantiles[len(quantiles)-1, i]
+            stats[prefix+"min_"+name]=quantiles[0, i]
+            for j in range(1, len(quantiles)-1):
+                stats[prefix+"quantile_"+name+str([0, 0.01, 0.05, 0.1, 0.25, 0.5, 0.75, 0.9, 0.95, 0.99, 1][j])]=quantiles[j, i]
+            for j in range(0, len(confidence_level)):
+                #stats[prefix+"confidence_level"+name+str(confidence_level[j])] = [mean[i]-stats.norm.ppf(confidence_level[j]/2)*variance_estimate[i], mean[i]+stats.norm.ppf(1-confidence_level[j]/2)*variance_estimate[i]] 
+                continue #something wrong with bootstrap
+            
+        if self.save_path!=None:
+            os.makedirs(self.save_path, exist_ok=True)
+            torch.save(stats, os.path.join(self.save_path,self.fn_cls.name))
+
+        return stats
+
+    def evaluate(self, models: Iterable[ContextModel]):# -> Iterable[Tensor]:
+
+        
+        sequence_length=self.fn_cls.sequence_length
+        batch_size=self.fn_cls.batch_size
+        errs=torch.zeros((len(models), self.num_batches, batch_size, sequence_length))
+
+        for i, (x_batch, y_batch) in zip(range(self.num_batches), self.fn_cls):
+            with torch.no_grad():
+                errs[:, i] = torch.stack([
+                  
+                    self._metric(
+                        y_batch,
+                        model.forward(x_batch, y_batch)
+                    )
+                    for model in models
+                ])
+        
+        errs=torch.reshape(errs, (len(models), self.num_batches*batch_size, sequence_length))
+
+        return self.PostProcessingStats(errs, [model.name for model in models]), errs
```

Comment:
This should also be run in the hosting scope -- e.g. the notebook will be returned the errs and fed in the models themselves, so they can call `PostProcessingStats` themselves.

Reply:
changed.